pinot-spi/src/main/java/org/apache/pinot/spi/utils/PinotReflectionUtils.java

```diff
@@ -19,36 +19,95 @@
+    try {
+      synchronized (REFLECTION_LOCK) {
+        return new Reflections(new ConfigurationBuilder().setUrls(ClasspathHelper.forPackage(packageName))
+            .filterInputsBy(new FilterBuilder.Include(regexPattern))
+            .setScanners(new TypeAnnotationsScanner())).getTypesAnnotatedWith(annotation);
+      }
+    } catch (Throwable t) {
+      // Log an error then re-throw it because this method is usually called in a static block, where exception might
+      // not be properly handled
+      LOGGER.error("Error scanning classes within package: {} with regex pattern: {}, annotation: {}", packageName,
+          regexPattern, annotation, t);
+      throw t;
+    }
+  }
+
+  public static Set<Class<?>> getClassesThroughReflection(List<String> packages, String regexPattern,
+      Class<? extends Annotation> annotation) {
+    try {
+      synchronized (REFLECTION_LOCK) {
+        List<URL> urls = new ArrayList<>();
+        for (String packageName : packages) {
+          urls.addAll(ClasspathHelper.forPackage(packageName));
+        }
+        return new Reflections(
+            new ConfigurationBuilder().setUrls(urls).filterInputsBy(new FilterBuilder.Include(regexPattern))
+                .setScanners(new TypeAnnotationsScanner())).getTypesAnnotatedWith(annotation);
+      }
+    } catch (Throwable t) {
+      // Log an error then re-throw it because this method is usually called in a static block, where exception might
+      // not be properly handled
+      LOGGER.error("Error scanning classes within packages: {} with regex pattern: {}, annotation: {}", packages,
+          regexPattern, annotation, t);
+      throw t;
+    }
+  }
+
+  public static Set<Method> getMethodsThroughReflection(String regexPattern, Class<? extends Annotation> annotation) {
+    return getMethodsThroughReflection(PINOT_PACKAGE_NAME, regexPattern, annotation);
+  }
+
+  public static Set<Method> getMethodsThroughReflection(String packageName, String regexPattern,
+      Class<? extends Annotation> annotation) {
+    try {
+      synchronized (REFLECTION_LOCK) {
+        return new Reflections(new ConfigurationBuilder().setUrls(ClasspathHelper.forPackage(packageName))
+            .filterInputsBy(new FilterBuilder.Include(regexPattern))
+            .setScanners(new MethodAnnotationsScanner())).getMethodsAnnotatedWith(annotation);
+      }
+    } catch (Throwable t) {
+      // Log an error then re-throw it because this method is usually called in a static block, where exception might
+      // not be properly handled
+      LOGGER.error("Error scanning methods within package: {} with regex pattern: {}, annotation: {}", packageName,
+          regexPattern, annotation, t);
+      throw t;
     }
   }
 
   /**
    * Due to the multi-threading issue in org.reflections.vfs.ZipDir, we need to put a lock before calling the
    * reflection related methods.
-   *
-   * @return
    */
   public static Object getReflectionLock() {
```

Comment:
This is no longer needed i think

Reply:
This is needed to wrap the setup of swagger. Added `runWithLock()` as @klsince suggested, and deprecated this one